sway/tree/view.c

```diff
@@ -689,18 +691,47 @@ static void subsurface_get_root_coords(struct sway_view_child *child,
 	}
 }
 
+static void subsurface_destroy(struct sway_view_child *child) {
+	if (!sway_assert(child->impl == &subsurface_impl,
+			"Expected a subsurface")) {
+		return;
+	}
+	struct sway_subsurface *subsurface = (struct sway_subsurface *)child;
+	wl_list_remove(&subsurface->destroy.link);
+	free(subsurface);
+}
+
 static const struct sway_view_child_impl subsurface_impl = {
 	.get_root_coords = subsurface_get_root_coords,
+	.destroy = subsurface_destroy,
 };
 
+static void view_child_damage(struct sway_view_child *child, bool whole);
+
+static void subsurface_handle_destroy(struct wl_listener *listener,
+		void *data) {
+	struct sway_subsurface *subsurface =
+		wl_container_of(listener, subsurface, destroy);
+	struct sway_view_child *child = &subsurface->child;
+	if (child->view->container != NULL) {
+		view_child_damage(child, true);
+	}
+	view_child_destroy(child);
```

Comment:
Doesn't this mean we call `view_child_destroy` twice now? `view_child_init` already sets up a listener on the surface's destroy event and calls it from the destroy handler.

Reply:
We have two objects: the `wl_surface` itself and the `wl_subsurface`. This listener is listening on the subsurface destroy event.

Destructors remove all listeners, so we should be fine if one is called before the other.